File: src/pi/modules/lib/kalman.py
class Kalman:

	def __init__(self, process_variance, measurement_variance, kalman_value):
		self.process_variance = process_variance ** 2
		self.measurement_variance = measurement_variance ** 2
		self.kalman_value = kalman_value
		self.sensor_value = kalman_value
		self.P, self.K = 1.0, 1.0
		self.kalman_value_list = []
		self.sensor_value_list = []

	def update_kalman(self, sensor_value):
		self.P += self.process_variance
		self.K = self.P / (self.P + self.measurement_variance)
		self.kalman_value = self.K * sensor_value + (1 - self.K) * self.kalman_value
		self.P *= (1 - self.K)

		self.kalman_value_list.append(self.kalman_value)
		self.sensor_value_list.append(sensor_value)
		return self.kalman_value

File: src/modules/lib/Kalman.cpp
//
// Created by adiv413 on 4/27/2020.
//

#include <flight/modules/lib/Kalman.hpp>

Kalman::Kalman(double process_variance, double measurement_variance, double kalman_value) {
    this->process_variance = process_variance * process_variance;
    this->measurement_variance = measurement_variance * measurement_variance;
    this->kalman_value = kalman_value;
    this->sensor_value = kalman_value;
    this->P = 1.0;
    this->K = 1.0;
}

double Kalman::update_kalman(double sensor_value_) {
    this->P += this->process_variance;
    this->K = this->P / (this->P + this->measurement_variance);
    this->kalman_value = this->K * sensor_value_ + (1 - this->K) * this->kalman_value;
    this->P *= (1 - this->K);

    return this->kalman_value;
}